flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAccumulator.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions.aggfunctions
+
+import java.lang.{Long => JLong}
+import java.lang.{Iterable => JIterable}
+import java.util.{Map => JMap}
+
+import org.apache.flink.table.api.dataview.MapView
+
+/**
+  * The base class for accumulator wrapper when applying distinct aggregation.
+  * @param realAcc the actual accumulator which gets invoke after distinct filter.
+  * @param mapView the [[MapView]] element used to store the distinct filter hash map.
+  * @tparam E the element type for the distinct filter hash map.
+  * @tparam ACC the accumulator type for the realAcc.
+  */
+class DistinctAccumulator[E, ACC](var realAcc: ACC, var mapView: MapView[E, JLong]) {
+  def this() {
+    this(null.asInstanceOf[ACC], new MapView[E, JLong]())
+  }
+
+  def this(realAcc: ACC) {
+    this(realAcc, new MapView[E, JLong]())
+  }
+
+  def getRealAcc: ACC = realAcc
+
+  def canEqual(a: Any): Boolean = a.isInstanceOf[DistinctAccumulator[E, ACC]]
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case that: DistinctAccumulator[E, ACC] => that.canEqual(this) &&
+        this.mapView == that.mapView
+      case _ => false
+    }
+
+  def add(element: E): Boolean = {
+    if (element != null) {
```

Comment:
Shouldn't it be up to the aggregation function how to handle `null` valued parameters?
I'd pass the null value on to the function instead of filtering it out. We'd need to wrap it in a `Row()` with a single field to distinguish the case of empty entry and null-valued entry.

Reply:
Great catch @fhueske. Sorry I misunderstood what you meant previous regrading `null` value. This line was to disregard all `null` values. But as you mentioned, some operators might treat `null` value differently, e.g. `COUNT` aggregators. I added the handling and the tests as well.